PROYECTO Sales-Prediction: Semana2

OBJETIVO: El proyecto consiste en realizar una predicción de ventas para productos alimenticios que se venden en distintas tiendas. El objetivo es ayudar a la tienda a comprender las propiedades de los productos y establecimientos que tienen un rol crucial en aumentar las ventas.

Para iniciar, importamos las librerarias a utilizar. Pandas y Numpy

In [1]:
import pandas as pd
import numpy as np

Una vez importadas las librerias, procedemos a dar lectura al dataframe

El objetivo de esta sección será:

- Asegurar de que se han cargado los datos correctamente.
- Ver qué tipo de datos tienes.
- Revisar la validez de los datos

In [2]:
df = pd.read_csv('/users/sebastianmoyacamus/codingdojo/sales_predictions.csv')

In [3]:
df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat         0.016047   
1           DRC01         5.92          Regular         0.019278   
2           FDN15        17.50          Low Fat         0.016760   
3           FDX07        19.20          Regular         0.000000   
4           NCD19         8.93          Low Fat         0.000000   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy  249.8092            OUT049   
1            Soft Drinks   48.2692            OUT018   
2                   Meat  141.6180            OUT049   
3  Fruits and Vegetables  182.0950            OUT010   
4              Household   53.8614            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1          3735.1380  
1  Supermarket Type2           443.4228  
2  Supermarket Type1          2097.2700  
3      Grocery Store           732.3800  
4  Supermarket Type1           994.7052

In [4]:
#df.apply(lambda x: sum(x.isnull()))

In [5]:
#df.apply(lambda x: len(x.unique()))

In [6]:
df.tail()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8518           FDF22        6.865          Low Fat         0.056783   
8519           FDS36        8.380          Regular         0.046982   
8520           NCJ29       10.600          Low Fat         0.035186   
8521           FDN46        7.210          Regular         0.145221   
8522           DRG01       14.800          Low Fat         0.044878   

               Item_Type  Item_MRP Outlet_Identifier  \
8518         Snack Foods  214.5218            OUT013   
8519        Baking Goods  108.1570            OUT045   
8520  Health and Hygiene   85.1224            OUT035   
8521         Snack Foods  103.1332            OUT018   
8522         Soft Drinks   75.4670            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
8518  Supermarket Type1          2778.3834  
8519  Supermarket Type1           549.2850  
8520  Supermarket Type1          1193.1136  
8521  Supermarket Type2          1845.5976  
8522  Supermarket Type1           765.6700

Luego, revisamos el tipo de datos de la columna
Queremos ver los tipos de datos de la columna para ver con qué tipo de datos estamos trabajando. Una vez que comencemos a construir modelos, necesitamos que todos nuestros datos estén en un tipo de columna numérica (enteros o flotantes) - así que presta más atención a cualquier columna de tipo «objeto».

In [7]:
# Verificar los tipos de datos de las columna usando el atributo dtypes
df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [8]:
# Usamos el atributo shape para obtener el número de filas y columnas del marco de datos
df.shape

(8523, 12)

Identificar datos faltantes

En Pandas, los valores faltantes se suelen llamar «NaN» o «None».

Los datos faltantes pueden:

- Indicar errores de recolección de datos.
- Indicar conversiones o manipulaciones inadecuadas.
- Los datos faltantes pueden estar listados como «cero», «falso», «no aplicable», «ingresados en una cadena vacía», entre otras posibilidades.

In [9]:
# Usamos el método info da a la columna tipos de datos + el número de valores no nulos
# De esta manera verificamos si hay valores perdidos
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


Con el método apply, podemos observar la cantidad de filas nulas para la columna Item Weight y Outlet Size

In [10]:
df.apply(lambda x: sum(x.isnull()))

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

Nuestro DF, está compuesto por 12 columnas y podemos observar que cuenta con valores nulos en las columnas "Item_Weight" y "Outlet_Size".

In [11]:
df.apply(lambda x: len(x.unique()))

Item_Identifier              1559
Item_Weight                   416
Item_Fat_Content                5
Item_Visibility              7880
Item_Type                      16
Item_MRP                     5938
Outlet_Identifier              10
Outlet_Establishment_Year       9
Outlet_Size                     4
Outlet_Location_Type            3
Outlet_Type                     4
Item_Outlet_Sales            3493
dtype: int64

Hay 1559 productos y 10 outlet. En item type podemos observar 16 valores únicos 

In [12]:
#Existe otro método para determinar las filas NaN
Weight_missing = df['Item_Weight'].isna()
# Mira las filas que contienen NaN para la columna "Item Weight"
df.loc[Weight_missing,:]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
18             DRI11          NaN          Low Fat         0.034238   
21             FDW12          NaN          Regular         0.035400   
23             FDC37          NaN          Low Fat         0.057557   
29             FDC14          NaN          Regular         0.072222   
...              ...          ...              ...              ...   
8485           DRK37          NaN          Low Fat         0.043792   
8487           DRG13          NaN          Low Fat         0.037006   
8488           NCN14          NaN          Low Fat         0.091473   
8490           FDU44          NaN          Regular         0.102296   
8504           NCN18          NaN          Low Fat         0.124111   

                  Item_Type  Item_MRP Outlet_Identifier  \
7               Snack Foods  107.7622            OUT027   
18              Hard Drinks  113.2834            OUT027   
21             Baking Goods  144.5444            OUT027   
23             Baking Goods  107.6938            OUT019   
29                   Canned   43.6454            OUT019   
...                     ...       ...               ...   
8485            Soft Drinks  189.0530            OUT027   
8487            Soft Drinks  164.7526            OUT027   
8488                 Others  184.6608            OUT027   
8490  Fruits and Vegetables  162.3552            OUT019   
8504              Household  111.7544            OUT027   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
7                          1985      Medium               Tier 3   
18                         1985      Medium               Tier 3   
21                         1985      Medium               Tier 3   
23                         1985       Small               Tier 1   
29                         1985       Small               Tier 1   
...                         ...         ...                  ...   
8485                       1985      Medium               Tier 3   
8487                       1985      Medium               Tier 3   
8488                       1985      Medium               Tier 3   
8490                       1985       Small               Tier 1   
8504                       1985      Medium               Tier 3   

            Outlet_Type  Item_Outlet_Sales  
7     Supermarket Type3          4022.7636  
18    Supermarket Type3          2303.6680  
21    Supermarket Type3          4064.0432  
23        Grocery Store           214.3876  
29        Grocery Store           125.8362  
...                 ...                ...  
8485  Supermarket Type3          6261.8490  
8487  Supermarket Type3          4111.3150  
8488  Supermarket Type3          2756.4120  
8490      Grocery Store           487.3656  
8504  Supermarket Type3          4138.6128  

[1463 rows x 12 columns]

In [13]:
#Sumamos todas las filas con valores perdidos para la columna Item_Weight
df['Item_Weight'].isna().sum()

1463

Observamos que, tenemos 1463 filas nulas para la columna Item_Weight.
Repetimos el proceso para la columna Outlet_Size

In [14]:
Size_missing = df['Outlet_Size'].isna()
# Observamos las filas que contienen NaN para la columna "Outlet_Size"
df.loc[Size_missing,:]

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
3              FDX07       19.200          Regular         0.000000   
8              FDH17       16.200          Regular         0.016687   
9              FDU28       19.200          Regular         0.094450   
25             NCD06       13.000          Low Fat         0.099887   
28             FDE51        5.925          Regular         0.161467   
...              ...          ...              ...              ...   
8502           NCH43        8.420          Low Fat         0.070712   
8508           FDW31       11.350          Regular         0.043246   
8509           FDG45        8.100          Low Fat         0.214306   
8514           FDA01       15.000          Regular         0.054489   
8519           FDS36        8.380          Regular         0.046982   

                  Item_Type  Item_MRP Outlet_Identifier  \
3     Fruits and Vegetables  182.0950            OUT010   
8              Frozen Foods   96.9726            OUT045   
9              Frozen Foods  187.8214            OUT017   
25                Household   45.9060            OUT017   
28                    Dairy   45.5086            OUT010   
...                     ...       ...               ...   
8502              Household  216.4192            OUT045   
8508  Fruits and Vegetables  199.4742            OUT045   
8509  Fruits and Vegetables  213.9902            OUT010   
8514                 Canned   57.5904            OUT045   
8519           Baking Goods  108.1570            OUT045   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
3                          1998         NaN               Tier 3   
8                          2002         NaN               Tier 2   
9                          2007         NaN               Tier 2   
25                         2007         NaN               Tier 2   
28                         1998         NaN               Tier 3   
...                         ...         ...                  ...   
8502                       2002         NaN               Tier 2   
8508                       2002         NaN               Tier 2   
8509                       1998         NaN               Tier 3   
8514                       2002         NaN               Tier 2   
8519                       2002         NaN               Tier 2   

            Outlet_Type  Item_Outlet_Sales  
3         Grocery Store           732.3800  
8     Supermarket Type1          1076.5986  
9     Supermarket Type1          4710.5350  
25    Supermarket Type1           838.9080  
28        Grocery Store           178.4344  
...                 ...                ...  
8502  Supermarket Type1          3020.0688  
8508  Supermarket Type1          2587.9646  
8509      Grocery Store           424.7804  
8514  Supermarket Type1           468.7232  
8519  Supermarket Type1           549.2850  

[2410 rows x 12 columns]

In [15]:
#Sumamos todas las filas con valores perdidos para la columna Outlet_Size
df['Outlet_Size'].isna().sum()

2410

Para la columna Outlet_Size, se observan 2.410 filas nulas.

In [16]:
df1 = df

In [17]:
filtro5 = df1['Item_Identifier'] == 'FDP10'
filtro6 = df1.loc[filtro5]
filtro6


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
7              FDP10          NaN          Low Fat         0.127470   
585            FDP10         19.0          Low Fat         0.128066   
2623           FDP10         19.0          Low Fat         0.128815   
3382           FDP10         19.0          Low Fat         0.128289   
4585           FDP10         19.0          Low Fat         0.128090   
6087           FDP10         19.0          Low Fat         0.128350   
7883           FDP10         19.0          Low Fat         0.127984   

        Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
7     Snack Foods  107.7622            OUT027                       1985   
585   Snack Foods  104.3622            OUT035                       2004   
2623  Snack Foods  107.6622            OUT017                       2007   
3382  Snack Foods  104.9622            OUT049                       1999   
4585  Snack Foods  107.0622            OUT046                       1997   
6087  Snack Foods  106.5622            OUT045                       2002   
7883  Snack Foods  107.6622            OUT013                       1987   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
7         Medium               Tier 3  Supermarket Type3          4022.7636  
585        Small               Tier 2  Supermarket Type1          1905.5196  
2623         NaN               Tier 2  Supermarket Type1          1164.4842  
3382      Medium               Tier 1  Supermarket Type1          1164.4842  
4585       Small               Tier 1  Supermarket Type1          1376.2086  
6087         NaN               Tier 2  Supermarket Type1          1482.0708  
7883        High               Tier 3  Supermarket Type1          1270.3464

In [18]:
filtro6.loc[7]

Item_Identifier                          FDP10
Item_Weight                                NaN
Item_Fat_Content                       Low Fat
Item_Visibility                        0.12747
Item_Type                          Snack Foods
Item_MRP                               107.762
Outlet_Identifier                       OUT027
Outlet_Establishment_Year                 1985
Outlet_Size                             Medium
Outlet_Location_Type                    Tier 3
Outlet_Type                  Supermarket Type3
Item_Outlet_Sales                      4022.76
Name: 7, dtype: object

Podemos observar, que cada identificador tiene un peso asignado. Por lo tanto, FDP10 

LIMPIEZA DE DATOS: METODOLOGÍA PROPUESTA

In [19]:
df.shape

(8523, 12)

In [20]:
df.apply(lambda x: sum(x.isnull()))

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [21]:
df2 = df

In [22]:
df2.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

In [23]:
item_avg_weight = df2.pivot_table(values='Item_Weight', index='Item_Identifier')
item_avg_weight

Item_Weight
Item_Identifier             
DRA12                 11.600
DRA24                 19.350
DRA59                  8.270
DRB01                  7.390
DRB13                  6.115
...                      ...
NCZ30                  6.590
NCZ41                 19.850
NCZ42                 10.500
NCZ53                  9.600
NCZ54                 14.650

[1555 rows x 1 columns]

In [32]:
df[:][df['Item_Identifier'] == 'DRI11']


Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
18             DRI11          NaN          Low Fat         0.034238   
1366           DRI11          NaN          Low Fat         0.060237   
3127           DRI11         8.26          Low Fat         0.034458   
5310           DRI11         8.26          Low Fat         0.057586   
7022           DRI11         8.26          Low Fat         0.034398   
7192           DRI11         8.26          Low Fat         0.034544   
8483           DRI11         8.26          Low Fat         0.034474   

        Item_Type  Item_MRP Outlet_Identifier  Outlet_Establishment_Year  \
18    Hard Drinks  113.2834            OUT027                       1985   
1366  Hard Drinks  115.9834            OUT019                       1985   
3127  Hard Drinks  113.3834            OUT049                       1999   
5310  Hard Drinks  113.7834            OUT010                       1998   
7022  Hard Drinks  115.7834            OUT035                       2004   
7192  Hard Drinks  116.0834            OUT018                       2009   
8483  Hard Drinks  117.0834            OUT045                       2002   

     Outlet_Size Outlet_Location_Type        Outlet_Type  Item_Outlet_Sales  
18        Medium               Tier 3  Supermarket Type3          2303.6680  
1366       Small               Tier 1      Grocery Store           345.5502  
3127      Medium               Tier 1  Supermarket Type1          2073.3012  
5310         NaN               Tier 3      Grocery Store           115.1834  
7022       Small               Tier 2  Supermarket Type1          2073.3012  
7192      Medium               Tier 3  Supermarket Type2          1267.0174  
8483         NaN               Tier 2  Supermarket Type1          1612.5676

Volvemos a comprobar que podemos recuperar de esta tabla el peso medio de todos los productos con el mismo Item_Identifier.

In [40]:
def impute_weight(cols):
    Weight = cols[0]
    Identifier = cols[1]
    
    if pd.isnull(Weight):
        return item_avg_weight['Item_Weight'][item_avg_weight.index == Identifier]
    else:
        return Weight
    
    
print('Orignal #missing: %d'%sum(df2['Item_Weight'].isnull()))
df2['Item_Weight'] = df2[['Item_Weight','Item_Identifier']].apply(impute_weight,axis=1).astype(float)
print('Final #missing: %d'%sum(data['Item_Weight'].isnull()))

Orignal #missing: 1463


ValueError: setting an array element with a sequence.